File: Core-backend-layer/Programming-fundamentals/Java/DesignPatterns/src/memento/ReservationHistory.java
package memento;
import java.util.Stack;

// Class to manage the history of reservations
public class ReservationHistory {
    private final Stack<ReservationMemento> history = new Stack<>();

    // Save a reservation state to the history
    public void saveReservation(ReservationMemento memento) {
        history.push(memento);
    }

    // Restore the most recent reservation state from the history
    public ReservationMemento undo() {
        if (!history.isEmpty()) {
            return history.pop();
        }
        return null;
    }
}

File: Core-backend-layer/Programming-fundamentals/Python/DesignPatterns/src/memento/reservation_history.py
# Class to manage the history of reservations
class ReservationHistory:
    def __init__(self):
        self.history = []

    # Save a reservation state to the history
    def save_reservation(self, reservation_memento):
        self.history.append(reservation_memento)

    # Restore the most recent reservation state from the history
    def undo(self):
        if self.history:
            return self.history.pop()
        return None